File: C++/rock_paper_scissors.cpp
#include <iostream>
#include <random>
#include <chrono>       //std::chrono::system_clock, std::chrono_literals
#include <thread>       // std::this_thread::sleep_for
#include <string>

using namespace std;

// does a 3 second countdown before each player's choice is revealed
void countdown()
{
    using namespace std::this_thread;     // sleep_for, sleep_until
    using namespace std::chrono_literals; // ns, us, ms, s, h, etc.
    using std::chrono::system_clock;
    
    for (int i = 3; i > 0; i--)
    {
        cout << i << endl;
        sleep_for(1s);
    }
    cout << endl << endl << endl;
}


// determines player choice of attack
string determineChoice(int person_play)
{
    if (person_play ==  1)
        return "Rock";
    else if (person_play == 2)
        return "Paper";
    else if (person_play == 3)
        return "Scissors";
}


//determines winner of each round
string determineWinner(int user_play, int bot_play)
{
    if (user_play == bot_play)
    {
        cout << "Tie. No one won." << endl;
        return "tie";
    }
    else if ((user_play == 1 && bot_play == 3) || (user_play == 2 && bot_play == 1) || (user_play == 3 && bot_play == 2))
    {
        cout << "You win!" << endl;
        return "player_win";
    }
    else
    {
        cout << "Awww you lost." << endl;
        return "bot_win";
    }
}


int main()
{
    cout << "Welcome to \"Rock, Papper, Scissors\". "
    << "You are playing against a bot and the best two out of three wins." << endl;
    
    bool someone_won = false; //determines if someone won the game
    int player_wins = 0; //no. of rounds player won
    int bot_wins = 0; //no. of rounds bot won
    int user_play, bot_play;

    //*************************************************//
    //********* RANDOM PLAY GENERATOR FOR BOT *********//
        int min = 1,
        max = 3;
        random_device rd;
        mt19937 bot_eng(rd());
        uniform_int_distribution <> uniformDist(min, max);
    //*************************************************//

    // Play more rounds until someone wins
    while (!someone_won)
    {
        cout << endl << "Press Enter key when ready..." << endl;
        cin.ignore(); // accepts a newline character

        cout << "Choose one of the following:\n1. Rock\n2. Paper\n3. Scissors" << endl
            << "Task ====> ";
        cin >> user_play;

        // While user input is invalid (e.g. string)
        while (cin.fail())
        {
            cin.clear(); // Clear error flags. This corrects the stream
            cin.ignore(1000, '\n'); // This skips the left over stream data
                                    // first param: arbitrarily large value
                                    // second param: character to ignore until in stream data  
            cout << "Please enter a number only." << endl
                << "Task ====> ";
            cin >> user_play;
        }

        bot_play = uniformDist(bot_eng); // Generate the bot's choice of attack

        countdown(); // Do a dramatic countdown

        // Reveal player attacks
        cout << "Player chose " << determineChoice(user_play) << endl;
        cout << "Bot chose " << determineChoice(bot_play) << endl;

        // Determine the winner of the round
        string round_winner = determineWinner(user_play, bot_play);

        if (round_winner == "player_win")
            player_wins++;
        else if (round_winner == "bot_win")
            bot_wins++;
        
        if (player_wins == 2 || bot_wins == 2)
            someone_won = true;
    }

    cout << endl << endl;
    if (player_wins == 2)
        cout << "You won the game!!! CONGRATULATIONS!!! ^_^" << endl;
    else
        cout << "The bot remains superior! Better luck next time." << endl;
}

File: PYTHON/rock_paper_scissors.py
import random
import time


class Game:
    def __init__(self, new_player_name):
        self.player_name = new_player_name
        self.player_hand = "NA"
        self.bot_hand = "NA"
        self.winner = "No winner yet"
        print("New instance of game class for " + self.player_name)

    def run_game(self):
        if self.player_hand == self.bot_hand:
            self.winner = "Tie. No one won."
        elif (self.player_hand == 'Rock' and self.bot_hand == 'Scissors') or (self.player_hand == 'Paper' and self.bot_hand == 'Rock') or (
                self.player_hand == 'Scissors' and self.bot_hand == 'Paper'):
            self.winner = "You win!"
        elif (self.player_hand == 'Rock' and self.bot_hand == 'Paper') or (self.player_hand == 'Paper' and self.bot_hand == 'Scissors') or (
                self.player_hand == 'Scissors' and self.bot_hand == 'Rock'):
            self.winner = "Awww you lost."


def countdown():
    for i in range(3, 0, -1):
        print(i)
        time.sleep(1)


def determine_choice(person_play):
    if person_play == 1:
        return "Rock"
    elif person_play == 2:
        return "Paper"
    elif person_play == 3:
        return "Scissors"


def determine_winner(user_play, bot_play):
    if user_play == bot_play:
        print("Tie. No one won.")
        return "tie"
    elif (user_play == 'Rock' and bot_play == 'Scissors') or (user_play == 'Paper' and bot_play == 'Rock') or (user_play == 'Scissors' and bot_play == 'Paper'):
        print("You win!")
        return "win"
    elif (user_play == 'Rock' and bot_play == 'Paper') or (user_play == 'Paper' and bot_play == 'Scissors') or (user_play == 'Scissors' and bot_play == 'Rock'):
        print("Awww you lost.")
        return "lose"


def main():
    print("Welcome to rock player scissors. You are playing against a bot and the best two out of three wins!")
    name = input("Can I ask of your name? ====> ")
    print("Pleased to meet you, " + name + ".")
    my_game = Game(name)
    print(my_game.player_name)
    print(my_game.player_hand)
    print(my_game.bot_hand)
    print(my_game.winner)

    my_game.player_hand = 'Paper'
    my_game.bot_hand = 'Scissors'
    my_game.run_game()

    print(my_game.winner)
    """
    print("Welcome to rock player scissors. You are playing against a bot and the best two out of three wins!")
    someone_won = False
    player_wins = 0
    bot_wins = 0
    while not someone_won:
        input("Press Enter key when ready...")

        while True:
            try:
                user_play = int(input("Choose one of the following:\n1. Rock\n2. Paper\n3. Scissors"
                                      "\nTask ====> "))
            except ValueError:
                print("Invalid input. Try again.")
                continue
            if user_play < 1 or user_play > 3:
                print("I don't understand. Please try again.")
            else:
                break
        bot_play = random.randint(1, 3)

        Countdown()
        print("Player chose", DetermineChoice(user_play))
        print("Bot chose", DetermineChoice(bot_play))
        round_winner = DetermineWinner(user_play, bot_play)
        if round_winner == "win":
            player_wins += 1
        elif round_winner == "lose":
            bot_wins += 1

        if player_wins == 2 or bot_wins == 2:
            someone_won = True

    print("\n\n")
    if player_wins == 2:
        print("You won the game!!! CONGRATULATIONS!!! ^_^")
    else:
        print("The bot remains superior! Better luck next time.")
    """


main()
